File: sdk-typescript/tests/conformance.test.ts
import assert from "node:assert/strict";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import test from "node:test";
import { z } from "zod";
import {
  Capsule,
  CapsuleGuardError,
  Destination,
  Policy,
  classified,
  classifyPayload,
  evaluatePolicy,
  zodClassified
} from "../src/index.ts";

const root = resolve("..");
const crmPolicy = JSON.parse(readFileSync(resolve(root, "fixtures/policies/crm-policy.json"), "utf8")) as Policy;
const conformance = JSON.parse(readFileSync(resolve(root, "fixtures/conformance/policy-decisions.json"), "utf8")) as {
  cases: Array<{
    name: string;
    policy: string;
    destination_id: string;
    destination_risk: "low" | "medium" | "high" | "critical";
    data_classes: string[];
    fields: string[];
    expected: { action: string; fields: string[]; reason: string };
  }>;
};

test("shared policy decision fixtures match TypeScript evaluator", () => {
  for (const item of conformance.cases) {
    const policy = JSON.parse(readFileSync(resolve(root, item.policy), "utf8")) as Policy;
    const decision = evaluatePolicy(
      policy,
      item.destination_id,
      item.destination_risk,
      item.data_classes,
      item.fields,
      "guard"
    );
    assert.equal(decision.action, item.expected.action, item.name);
    assert.deepEqual(decision.fields, item.expected.fields, item.name);
    assert.equal(decision.reason, item.expected.reason, item.name);
  }
});

test("zod field annotations participate in classification", () => {
  const emailSchema = zodClassified(z.string().email(), ["email"]);
  const parsed = emailSchema.parse("claimant@example.com");
  assert.deepEqual(classifyPayload({ email: parsed, support_tier: "gold" }), ["email", "support_tier"]);
});

test("async context propagates through wrapped tool calls", async () => {
  const destination: Destination = {
    id: "crm",
    type: "external_tool",
    domain: "api.crm.example",
    provider: "Example CRM",
    risk: "high"
  };
  const received: unknown[] = [];
  const capsule = new Capsule({ mode: "guard", policy: crmPolicy, agentName: "claims-triage" });
  const tool = capsule.wrapTool("crm.upsert_account", destination, async (payload: unknown) => {
    await Promise.resolve();
    received.push(payload);
    return { ok: true };
  });

  await capsule.run("phase13-typescript", async () => {
    await Promise.all([
      tool({
        account_id: classified("acct_123", ["account_id"]),
        email: classified("claimant@example.com", ["email"]),
        account_notes: classified("Sensitive account note", ["account_notes"])
      }),
      tool({
        account_id: classified("acct_456", ["account_id"]),
        email: classified("second@example.com", ["email"]),
        account_notes: classified("Second sensitive note", ["account_notes"])
      })
    ]);
  });

  assert.equal(received.length, 2);
  assert.match(JSON.stringify(received), /\[redacted:email\]/);
  const trace = capsule.trace();
  assert.equal(trace.language, "typescript");
  assert.equal(trace.spans.length, 2);
  assert.deepEqual(new Set(trace.spans.map((span) => span.policy_decision.action)), new Set(["redact"]));
  assert.equal(trace.destinations[0]?.declared_in_policy, true);

  const output = resolve("build/conformance/typescript-trace.json");
  mkdirSync(dirname(output), { recursive: true });
  writeFileSync(output, `${JSON.stringify(trace, null, 2)}\n`, "utf8");
  assert.doesNotMatch(JSON.stringify(trace), /claimant@example.com|Sensitive account note/);
});

test("guard mode blocks undeclared high-risk egress", async () => {
  const restrictive = JSON.parse(readFileSync(resolve(root, "fixtures/policies/restrictive-policy.json"), "utf8")) as Policy;
  const capsule = new Capsule({ mode: "guard", policy: restrictive, agentName: "claims-triage" });
  const tool = capsule.wrapTool("external_ocr.extract", {
    id: "external_ocr",
    type: "external_tool",
    domain: "api.ocr.example",
    provider: "Example OCR",
    risk: "high"
  }, async (_payload: unknown) => ({ ok: true }));

  await assert.rejects(
    () => capsule.run("blocked-typescript", () => tool({
      document: classified("private document", ["document_text"]),
      medical_information: classified("private diagnosis", ["medical_information"])
    })),
    CapsuleGuardError
  );
  assert.equal(capsule.trace().spans[0]?.status, "blocked");
});

File: sdk-java/src/test/java/dev/agentcapsule/ConformanceTest.java
package dev.agentcapsule;

import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public final class ConformanceTest {
    public static void main(String[] args) throws Exception {
        policyDecisionFixturesMatch();
        annotationClassificationWorks();
        builderAndInterceptorsPropagateContext();
        guardModeBlocksHighRiskEgress();
    }

    private static void policyDecisionFixturesMatch() {
        List<Case> cases = List.of(
                new Case(Policy.claimsTriagePolicy(), "crm", "high", List.of("email", "account_notes"), List.of("email", "account_notes"), "redact", List.of("account_notes", "email"), "destination redaction rule matched"),
                new Case(Policy.claimsTriagePolicy(), "crm", "high", List.of("account_id", "support_tier"), List.of("account_id", "support_tier"), "allow", List.of("account_id", "support_tier"), "destination declared and data allowed"),
                new Case(Policy.claimsTriagePolicy(), "crm", "high", List.of("medical_information"), List.of("medical_information"), "require_approval", List.of("medical_information"), "destination approval rule matched"),
                new Case(Policy.claimsTriagePolicy(), "crm", "high", List.of("account_id", "support_tier", "customer_identifier"), List.of("account_id", "support_tier", "customer_identifier"), "allow_fields", List.of("account_id", "support_tier"), "destination allowlist excluded fields"),
                new Case(Policy.restrictivePolicy(), "external_ocr", "high", List.of("medical_information", "document_text"), List.of("medical_information", "document_text"), "block", List.of("document_text", "medical_information"), "undeclared high-risk egress"),
                new Case(Policy.restrictivePolicy(), "analytics", "medium", List.of("support_tier"), List.of("support_tier"), "warn", List.of("support_tier"), "undeclared destination"),
                new Case(Policy.claimsTriagePolicy(), "crm", "high", List.of("secrets"), List.of("api_key"), "block", List.of("api_key"), "secrets default rule matched")
        );

        for (Case item : cases) {
            PolicyDecision decision = PolicyEngine.evaluate(item.policy, item.destinationId, item.destinationRisk, item.dataClasses, item.fields, "guard");
            require(decision.action().equals(item.action), "action mismatch for " + item.destinationId);
            require(decision.fields().equals(item.expectedFields), "fields mismatch for " + item.destinationId + ": " + decision.fields());
            require(decision.reason().equals(item.reason), "reason mismatch for " + item.destinationId);
        }
    }

    private static void annotationClassificationWorks() {
        ClaimPayload payload = new ClaimPayload("claimant@example.com", "Sensitive account note", "gold");
        require(FieldClassifier.classify(payload).equals(List.of("account_notes", "email", "support_tier")), "annotation classifier drifted");
    }

    private static void builderAndInterceptorsPropagateContext() throws Exception {
        Capsule capsule = Capsule.builder()
                .mode("guard")
                .policy(Policy.claimsTriagePolicy())
                .agentName("claims-triage")
                .build();
        Destination crm = new Destination("crm", "external_tool", "api.crm.example", "Example CRM", "high");
        Capsule.ThrowingFunction<ClaimPayload, Map<String, Object>> tool = capsule.interceptHttp(
                "crm.upsert_account",
                crm,
                payload -> Map.of("ok", true)
        );

        capsule.run("phase13-java", () -> {
            Capsule.ThrowingSupplier<Map<String, Object>> propagated = capsule.propagate(() ->
                    tool.apply(new ClaimPayload("claimant@example.com", "Sensitive note", "gold"))
            );
            CompletableFuture<Map<String, Object>> result = CompletableFuture.supplyAsync(() -> {
                try {
                    return propagated.get();
                } catch (Exception exc) {
                    throw new RuntimeException(exc);
                }
            });
            return result.join();
        });

        Capsule.Trace trace = capsule.trace();
        require(trace.language().equals("java"), "trace language mismatch");
        require(trace.spans().size() == 1, "span count mismatch");
        require(trace.spans().get(0).policyDecision().action().equals("redact"), "interceptor did not redact");
    }

    private static void guardModeBlocksHighRiskEgress() throws Exception {
        Capsule capsule = Capsule.builder()
                .mode("guard")
                .policy(Policy.restrictivePolicy())
                .agentName("claims-triage")
                .build();
        Capsule.ThrowingFunction<Map<String, Object>, Map<String, Object>> tool = capsule.wrapTool(
                "external_ocr.extract",
                new Destination("external_ocr", "external_tool", "api.ocr.example", "Example OCR", "high"),
                payload -> Map.of("ok", true)
        );

        boolean blocked = false;
        try {
            capsule.run("blocked-java", () -> tool.apply(Map.of(
                    "document", "private document",
                    "medical_information", "private diagnosis"
            )));
        } catch (Capsule.CapsuleGuardException exc) {
            blocked = true;
        }
        require(blocked, "guard mode did not block undeclared high-risk egress");
        require(capsule.trace().spans().get(0).status().equals("blocked"), "blocked span not recorded");
    }

    private static void require(boolean condition, String message) {
        if (!condition) {
            throw new AssertionError(message);
        }
    }

    private record Case(
            Policy policy,
            String destinationId,
            String destinationRisk,
            List<String> dataClasses,
            List<String> fields,
            String action,
            List<String> expectedFields,
            String reason
    ) {
    }

    private record ClaimPayload(
            @DataClass("email") String email,
            @DataClass("account_notes") String notes,
            @DataClass("support_tier") String supportTier
    ) {
    }
}
